src/TSTransformer/util/types.ts

```diff
@@ -20,21 +22,85 @@ function getRecursiveBaseTypes(type: ts.InterfaceType) {
+			return type.some(t => isDefinitelyTypeInner(state, t, callbacks, requireFullMatchNode));
+		}
+		// For any checks:
+		// With an array, *at least one* of the types must match *at least one* of the callbacks
+		// If this is not the case, *and* at least one of the types is `any`, report the error
+		// Intentionally omit the recursive `requireFullMatchNode` here in case one of the types matches
+		// In which case we don't want the noAny diagnostic to be reported
+		const numberOfFits = type.reduce((acc, t) => acc + (isDefinitelyTypeInner(state, t, callbacks) ? 1 : 0), 0);
+		if (numberOfFits === 0) {
+			// None of the types matched, so run with `requireFullMatchNode` to report possible noAny
+			type.forEach(t => isDefinitelyTypeInner(state, t, callbacks, requireFullMatchNode));
+		}
+		return numberOfFits > 0;
+	} else if (type.isUnion()) {
+		const numberOfFits = type.types.reduce(
+			(acc, t) => acc + (isDefinitelyTypeInner(state, t, callbacks, requireFullMatchNode) ? 1 : 0),
+			0,
+		);
+		// In a union, either *all* or *none* of the types should match
+		if (requireFullMatchNode && numberOfFits !== 0 && numberOfFits !== type.types.length) {
+			addDiagnosticFromNodeIfNotCached(
+				state,
+				requireFullMatchNode,
+				errors.noMixedTypes(requireFullMatchNode),
+				state.multiTransformState.isReportedByNoMultiCache,
+			);
+		}
+		return numberOfFits === type.types.length;
 	} else if (type.isIntersection()) {
-		return type.types.some(t => isDefinitelyTypeInner(t, callbacks));
+		return type.types.some(t => isDefinitelyTypeInner(state, t, callbacks, requireFullMatchNode));
 	} else {
-		if (type.isClassOrInterface() && getRecursiveBaseTypes(type).some(t => isDefinitelyTypeInner(t, callbacks))) {
+		if (requireFullMatchNode && isAnyType(type.checker)(type)) {
+			// `any` means the type does not definitely fit.
+			// However, we return true here,
+			// because some exhaustive type checks will assertion crash
+			// if they do not match any of the specific cases.
+			// For example, `getAddIterableToArrayBuilder`.
+			// Since a diagnostic is reported,
+			// it won't matter that the emitted code is wrong
+			addDiagnosticFromNodeIfNotCached(
+				state,
+				requireFullMatchNode,
+				errors.noAny(requireFullMatchNode),
+				state.multiTransformState.isReportedByNoAnyCache,
+			);
+			return true;
+		} else if (
+			type.isClassOrInterface() &&
+			getRecursiveBaseTypes(type).some(t => isDefinitelyTypeInner(state, t, callbacks, requireFullMatchNode))
+		) {
 			return true;
 		}
 		return callbacks.some(cb => cb(type));
 	}
 }
 
-export function isDefinitelyType(type: ts.Type, ...callbacks: Array<TypeCheck>) {
-	return isDefinitelyTypeInner(type.getConstraint() ?? type, callbacks);
+/** Returns true if the type definitely fits *at least one* of the provided callbacks */
+export function isDefinitelyType(
+	state: TransformState,
+	type: ts.Type | Array<ts.Type>,
+	requireFullMatchNode?: ts.Node,
+	...callbacks: Array<TypeCheck>
```

Comment:
If we're adding more arguments to this, we should definitely remove the `...` to make callsites clearer.

Reply:
If that would be an array, then what's wrong with the types allowing an array? That seems equally less readable.